index.js

```diff
@@ -651,6 +855,45 @@
     );
   };
 
+  var conflictingTypeVarReturnValue = function(name, type, value, prevTypes) {
+    var prevIndexes = prevTypes[0].indexes;
+    var argsStr = prevIndexes.length > 1 ? 'arguments' : 'argument';
+    return new TypeError(
+      LEFT_SINGLE_QUOTATION_MARK + name + RIGHT_SINGLE_QUOTATION_MARK +
+      ' expected its ' + ordinalList(prevIndexes) + ' ' + argsStr + ' and' +
+      ' return value to be of the same type; the ' + ordinalList(prevIndexes) +
+      ' ' + argsStr + ' ' + isAre(prevIndexes) + ' of type ' +
+      prevTypes.join(' or ') + ', however the return value, ' + show(value) +
+      ', is of type ' + type
+    );
+  };
+  
+  var ul = function(items) {
+    return map(items, prefix('  - ')).join('\n');
+  };
+
+  var conflictingTypeVarArgument = function(sig, name, index, value1, types1, value2, types2) {
+    return new TypeError(
+      'TypeVariable constraint violation\n\n' +
+      show(sig) + '\n' +
+      sigShowTypeVarPositions(sig, name) + '\n\n' +
+      'The value at position 1 is ' + show(value1) + ', which is a ' +
+      'member of the following types:\n\n' +
+      ul(types1)+'\n\n' +
+      'The value at position 2 is ' + show(value2) + ', which is a ' +
+      'member of the following types:\n\n' +
+      ul(types2) + '\n\n' +
+      'Since there is no type of which both values are members, the ' +
+      'type-variable constraint has been violated.\n'
+    );
+  };
+
+  var conflictingTypeVar = function(sig, name, index, value1, types1, value2, types2) {
+    return isNaN(index) ?
+      conflictingTypeVarReturnValue(sig, type, value1, prevTypes) :
+      conflictingTypeVarArgument(sig, name, index, value1, types1, value2, types2);
+  };
```

Comment:
I think we could use the same approach for both arguments and return values.

```
fromMaybe :: a -> Maybe a -> a
             ^          ^
             1          2
```

```
fromMaybe :: a -> Maybe a -> a
             ^               ^
             1               2
```

I'm guessing you haven't yet had a chance to work on the handling of invalid return values.